internal/plugin/invoke.go

```diff
@@ -0,0 +1,203 @@
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/hcl/v2/json"
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint-plugin-sdk/internal-shared/protomappers"
+)
+
+// PluginRequest describes a plugin that should be setup to have it's functions
+// invoked. Config is used to identify the plugin by name, and Type is used to identify
+// which one of the plugin types should be addressed within the plugin process.
+type PluginRequest struct {
+	// Config contains the information about the plugin itself. This will
+	// be used to locate the plugin so it can be started.
+	Config
+
+	// The different components
+	Type component.Type
+	Name string
+
+	Dir string
+
+	ConfigData []byte
+	JsonConfig bool
+}
+
+func startInstance(
+	ctx context.Context, log hclog.Logger, req *PluginRequest,
+) (*Instance, error) {
+	// Figure out the command to execute to start to bring the plugin
+	// into existance
+	pluginPaths, err := DefaultPaths(req.Dir)
+	if err != nil {
+		return nil, err
+	}
+
+	var fn interface{}
+
+	cmd, err := Discover(&req.Config, pluginPaths)
+	if err != nil {
+		return nil, err
+	}
+
+	if cmd != nil {
+		fn = Factory(cmd, req.Type)
+	} else {
+		// If the plugin was not found, it is only an error if
+		// we don't have that plugin as a builtin plugin (for builtin plugins
+		// we relaunch the existing process)
+
+		if _, ok := Builtins[req.Config.Name]; !ok {
+			return nil, fmt.Errorf("plugin %q not found", req.Config.Name)
+		} else {
+			log.Debug("plugin found as builtin")
+			fn = BuiltinFactory(req.Config.Name, req.Type)
+		}
+	}
+
+	// Use argmapper to call our plugin creation function with
+	// the standard context and log args. This is just to create
+	// the plugin process via go-plugin, not to actually invoke
+	// anything within the plugin yet.
+
+	afn, err := argmapper.NewFunc(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	fnResult := afn.Call(argmapper.Typed(ctx, log))
+	if err := fnResult.Err(); err != nil {
+		return nil, err
+	}
+
+	rawComponent := fnResult.Out(0)
+
+	// If we have a plugin.Instance then we can extract other information
+	// from this plugin. We accept pure factories too that don't return
+	// this so we type-check here. This would only happen if the above code
+	// used a creation function other than the *Factory functions.
+	pi, ok := rawComponent.(*Instance)
+	if !ok {
+		pi = &Instance{
+			Component: rawComponent,
+			Close:     func() {},
+		}
+	}
+
+	return pi, nil
+}
+
+// Plugin is the state of a created plugin to be invoked, returned by OpenPlugin.
+type Plugin struct {
+	req      *PluginRequest
+	Instance *Instance
+}
+
+// OpenPlugin resolves the plugin information in req and returns a Plugin value
+// to have Invoke called upon it. OpenPlugin also returns the raw component
+// interface, which is used to get the function value that will be invoked
+func OpenPlugin(
```

Comment:
Maybe just name this `Open`, since this is in the plugin package already: `plugin.Open` vs `plugin.OpenPlugin` the latter feels redundant.

Reply:
Totes.